In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
path = 'originalData'

In [4]:
#Import the “products.csv” file into Jupyter
products_df = pd.read_csv(os.path.join(path, 'products.csv'))

In [5]:
path = 'PreparedData'

In [8]:
df_ords = pd.read_csv(os.path.join(path,'orders_wrangled.csv'))

In [9]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [10]:
# create a dataframe

df_test = pd.DataFrame()

In [11]:
# create a mixed type column

df_test['mix'] = ['a','b',1,True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


/var/folders/n3/cnsw2_q5653501p8bw3zft6h0000gp/T/ipykernel_91550/1786467950.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [15]:
df_test['mix'] = df_test['mix'].astype('str')


In [21]:
products_df.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [22]:
df_nan = products_df[products_df['product_name'].isnull() == True]

In [23]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [24]:
products_df.shape

(49693, 5)

In [26]:
products_df_clean = products_df[products_df['product_name'].isnull() == False]

In [31]:
products_df_clean.shape

(49677, 5)

In [28]:
products_df.dropna(subset = ['product_name'], inplace = True)

In [29]:
df_dups = products_df_clean[products_df_clean.duplicated()] 

In [30]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [32]:
products_df_clean.shape

(49677, 5)

In [33]:
products_df_clean_no_dups = products_df_clean.drop_duplicates()

In [34]:
products_df_clean_no_dups.shape

(49672, 5)

In [36]:
new_path = 'PreparedData'

In [37]:
products_df_clean_no_dups.to_csv(os.path.join(new_path,'products_checked.csv'))

In [43]:
#Import the “orders_wrangled.csv” file into Jupyter
orders_df = pd.read_csv(os.path.join(path, 'orders_wrangled.csv'))



Excercise 4.5 Data Consistency Checks

# 2 Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [44]:
orders_df.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Depends on the data above most of the numbers seem normal, but we should check a few things. One column might have some missing data, so we need to figure out why. Another column shows the average time for orders on a day of the week, but the number seems a bit odd. The time when orders are placed varies a lot in one column, so we should look into why that happens. Also, make sure the numbers make sense in the real world, like the hours in a day. So, in short, we should investigate these things to make sure the data is reliable and tells us the right information.


# 3 Check for mixed-type data in your df_ords dataframe.

In [45]:
orders_df.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [51]:
for col in orders_df.columns.tolist():
  weird = (orders_df[[col]].applymap(type) != orders_df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (orders_df[weird]) > 0:
    print (col)

/var/folders/n3/cnsw2_q5653501p8bw3zft6h0000gp/T/ipykernel_91550/568139757.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (orders_df[[col]].applymap(type) != orders_df[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/n3/cnsw2_q5653501p8bw3zft6h0000gp/T/ipykernel_91550/568139757.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (orders_df[[col]].applymap(type) != orders_df[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/n3/cnsw2_q5653501p8bw3zft6h0000gp/T/ipykernel_91550/568139757.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (orders_df[[col]].applymap(type) != orders_df[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/n3/cnsw2_q5653501p8bw3zft6h0000gp/T/ipykernel_91550/568139757.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (orders_df[[col]].applymap(type) != orders_d

# 4 If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

The column in question should contain observations of a single data type: No mixed-type data found.

#5 Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [54]:
orders_df.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [55]:
orders_df.head(20)

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0            0   2539329        1    prior             1                   2   
1            1   2398795        1    prior             2                   3   
2            2    473747        1    prior             3                   3   
3            3   2254736        1    prior             4                   4   
4            4    431534        1    prior             5                   4   
5            5   3367565        1    prior             6                   2   
6            6    550135        1    prior             7                   1   
7            7   3108588        1    prior             8                   1   
8            8   2295261        1    prior             9                   1   
9            9   2550362        1    prior            10                   4   
10          10   1187899        1    train            11                   4   
11          11   2168274        2    prior             1                   2   
12          12   1501582        2    prior             2                   5   
13          13   1901567        2    prior             3                   1   
14          14    738281        2    prior             4                   2   
15          15   1673511        2    prior             5                   3   
16          16   1199898        2    prior             6                   2   
17          17   3194192        2    prior             7                   2   
18          18    788338        2    prior             8                   1   
19          19   1718559        2    prior             9                   2   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     NaN  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0  
15                 11                     8.0  
16                  9                    13.0  
17                 12                    14.0  
18                 15                    27.0  
19                  9                     8.0

In [56]:
# create a dataframe for the missing values
orders_df_nan = orders_df [orders_df ['days_since_prior_order'].isnull () == True]

In [57]:
orders_df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

The maximum value in the column 'user_id' is 206209, which corresponds with the number of missing values in the column 'days_since_prior_order' that represents one for each new customer. The reason for missing value is majority of customers are first time clients who never placed an order before.

# 6. Address the missing values using an appropriate method:

I’ve chosen to make no changes. . A value of "1’ in the column 'order_number' can represent new customers.This is because I believe that it isn’t reasonable to assign values or remove or filter the data. Statistics may still be analyzed with the NaN values present

#7 Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [58]:
orders_df_dups = orders_df[orders_df.duplicated()]

In [59]:
orders_df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No Duplicates Found

#8 Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

No duplicates Found. I have chosen to make no changes.

#9 Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.


In [61]:
new_path = 'PreparedData'

In [62]:
orders_df_dups.to_csv(os.path.join(new_path,'orders_checked.csv'))